server/app/services/applicant/ApplicantService.java

```diff
@@ -629,32 +646,110 @@ private ApplicationPrograms relevantProgramsForApplicant(
-            submittedPrograms.add(
+
+            Application submittedApp = maybeSubmittedApp.get();
+            ApplicantProgramData.Builder applicantProgramDataBuilder =
                 ApplicantProgramData.builder()
                     .setProgram(activeProgramNames.get(programName))
                     .setLatestSubmittedApplicationTime(latestSubmittedApplicationTime)
-                    .setLatestSubmittedApplicationStatus(maybeCurrentStatus)
-                    .build());
-            programNamesWithApplications.add(programName);
+                    .setLatestSubmittedApplicationStatus(maybeCurrentStatus);
+            programNamesWithApplicationsFuture.add(programNameFuture);
+
+            setApplicationsForType(
+                    submittedApp.getApplicant().id,
+                    submittedApp.getProgram().getProgramDefinition(),
+                    submittedPrograms,
+                    applicantProgramDataBuilder)
+                .thenApplyAsync(
+                    (v) -> {
+                      programNamesWithApplications.add(programName);
+                      return programNameFuture.complete(programName);
+                    })
+                .exceptionally(
+                    ex -> {
+                      throw new RuntimeException(ex);
+                    });
           }
         });
 
-    Set<String> unappliedActivePrograms =
-        Sets.difference(activeProgramNames.keySet(), programNamesWithApplications);
-    unappliedActivePrograms.forEach(
-        programName -> {
-          unappliedPrograms.add(
-              ApplicantProgramData.builder()
-                  .setProgram(activeProgramNames.get(programName))
-                  .build());
-        });
-
-    // Ensure each list is ordered by database ID for consistent ordering.
-    return ApplicationPrograms.builder()
-        .setInProgress(sortByProgramId(inProgressPrograms.build()))
-        .setSubmitted(sortByProgramId(submittedPrograms.build()))
-        .setUnapplied(sortByProgramId(unappliedPrograms.build()))
-        .build();
+    CompletableFuture<ApplicationPrograms> applicationProgramsFuture = new CompletableFuture<>();
+    if (mostRecentApplicationsByProgram.isEmpty()) {
+      Set<String> unappliedActivePrograms =
+          Sets.difference(activeProgramNames.keySet(), programNamesWithApplications);
+      unappliedActivePrograms.forEach(
+          programName -> {
+            unappliedPrograms.add(
+                ApplicantProgramData.builder()
+                    .setProgram(activeProgramNames.get(programName))
+                    .build());
+          });
+
+      applicationProgramsFuture.complete(
+          ApplicationPrograms.builder()
+              .setInProgress(sortByProgramId(inProgressPrograms.build()))
+              .setSubmitted(sortByProgramId(submittedPrograms.build()))
+              .setUnapplied(sortByProgramId(unappliedPrograms.build()))
+              .build());
+    } else {
```

Comment:
**suggestion (blocking):** the if/else branches in this method have become pretty big making it hard to follow the flow of the code, if feasible please factor code into smaller methods

Reply:
This is no longer applicable after refactoring and loading the data upfront.